tracing support

vyzo opened the pull request:
~Currently only the tracing scaffolding, but this should be enough for initial review pass.~

TBD:
- [x] implement tracer methods with particular event types
- [x] implement tracing to file
- [x] implement tracing to peer

vyzo committed:
tracing scaffolding

vyzo committed:
trace publish

vyzo committed:
add tracing to floodsub/randomsub

vyzo committed:
remove useless nil check when initializing subsystem tracers

vyzo committed:
initialize tracer with peer ID, trace RPC from join/leave announcements

vyzo committed:
trace event protobuf

vyzo committed:
some minor fixes in trace pb

vyzo committed:
implement tracing details

vyzo committed:
add traces for send/drop rpc

forgotten!

vyzo committed:
track topics in message tracing

vyzo committed:
json tracer

vyzo committed:
move tracer implementation to its own file

vyzo committed:
add protobuf file tracer

Stebalien reviewed (dismissed):
This seems too "not invented here".

Stebalien commented:
(we discussed this on a call but I misunderstood some important context)

If we were trying to trace across machines, I'd understand maybe doing this ourselves. Given that we're _just_ tracing events locally, this seems way over the top. Why not just use Zap?

If the answer is "performance", have you tested that? I really doubt logging is going to be a bottleneck given encrypted transports, signing, verifying, etc.

Stebalien commented:
Unless the concern is bandwidth. That's the one thing this buys us over a general-purpose logger like zap.

Stebalien commented on trace.go:
```diff
@@ -0,0 +1,336 @@
+
+// Generic event tracer interface
+type EventTracer interface {
+	Trace(evt interface{})
```
`pb.Message`?

vyzo replied:
Yeah, we could do that to avoid the cast.

vyzo replied:
done, it takes a `*pb.TraceEvent` now.

Stebalien commented:
Ok, I've discussed this with raul and looked at it a bit more. This still feels like a lot of code for something so simple but I understand why we're using protobufs (gRPC & bandwidth).

vyzo commented:
Yes, we _are_ trying to trace across machines, it's just that only the local tracer has been written so far.
Tracing to peer (and gRPC based tracing) are forthcoming.
Also note, bandwidth is a real concern here.

vyzo committed:
use *pb.TraceEvent as argument for Trace in the EventTracer interface

Stebalien commented:
> Yes, we are trying to trace across machines

I'm talking about including trace IDs in messages. However, message IDs should give us that anyways.

vyzo commented:
> I'm talking about including trace IDs in messages. 

The peer ID of the originating peer is included in the message.

vyzo committed:
remote tracer

vyzo committed:
remote tracer: wait a second to accumulate batches

vyzo commented:
Rebased on master.

vyzo committed:
remove CompressedTraceEventBatch from trace protobuf

vyzo committed:
compress entire stream in remote tracer

vyzo committed:
reset remote tracer stream on errors

Stebalien commented on floodsub.go:
```diff
@@ -39,11 +40,16 @@ func (fs *FloodSubRouter) Protocols() []protocol.ID {
 
-func (fs *FloodSubRouter) AddPeer(peer.ID, protocol.ID) {}
+func (fs *FloodSubRouter) AddPeer(p peer.ID, proto protocol.ID) {
+	fs.tracer.AddPeer(p, proto)
```
Should this be in the _router_ or in the part that calls `AddPeer`?

vyzo replied:
We could have it in the control substrate, it just felt more natural to do it in the router.
Other than that there is no particular reason for the choice.

Stebalien commented on floodsub.go:
```diff
@@ -39,11 +40,16 @@ func (fs *FloodSubRouter) Protocols() []protocol.ID {
 
-func (fs *FloodSubRouter) RemovePeer(peer.ID) {}
+func (fs *FloodSubRouter) RemovePeer(p peer.ID) {
+	fs.tracer.RemovePeer(p)
```
ditto.

vyzo replied:
same, it could be in either place, just felt more natural to do it in the router.

Stebalien commented on tracer.go:
```diff
@@ -0,0 +1,290 @@
+package pubsub
```
Double checking, this is moving to a new repo?

vyzo replied:
I think we can keep this file here, so that we have a one-stop construction of the pubsub system without requiring an external dependency.

Stebalien reviewed (commented):
After my initial "why are we doing this" reaction, this LGTM (modulo a few questions/nits).

vyzo committed:
lossy tracing for remote tracers

Don't accumulate memory if the tracer is being slow or unavailable, just drop the trace and log.

Stebalien commented on tracer.go:
Let's make this a bit safer.

vyzo replied:
Sure.

vyzo replied:
I added a closed state variable, which gets set when we first Close, and is also consulted before adding to the event buffer.
It should be safe and idempotent now, no potential panics because of a closed channel.

Stebalien commented on tracer.go:
We could end up with a pretty large buffer this way and/or drop messages. Is there a better way to do this?

vyzo replied:
We could check the size of the buffer and sleep only if the batch is smaller than a threshold.
We can also reduce the sleep time and poll a few times, say up to a second.

vyzo replied:
I added a check for the size of the batch, and polling every 100ms (for up to 1s) in order to accumulate buffer in safer way.

Stebalien commented on tracer.go:
Shouldn't we do this at this at the end?

vyzo replied:
that's the buffer we are swapping in, it's already been consumed.

Stebalien replied:
Yeah, but we could nil it out earlier to avoid holding on to the memory.

vyzo replied:
Oh, that's a good point. Ok, I'll move the niling out right after we write it.

vyzo replied:
done.

Stebalien commented on tracer.go:
Can't we just call open a stream? We shouldn't have to dial separately.

vyzo replied:
We can, but the host will have to know how to reach the trace peer. The traced is not announcing itself to the dht, so it wouldn't be findable in general.

Stebalien replied:
Ah. Shouldn't we add the the addresses to the peerstore on start with an expiration of "never" (https://github.com/libp2p/go-libp2p-core/blob/master/peerstore/peerstore.go#L45)?

vyzo replied:
that would work too; can do.

vyzo replied:
done.

vyzo committed:
move niling of trace buffer to the end

this avoids holding on memory while we are waiting.

vyzo committed:
don't blanket wait for 1s to accumulate a batch.

Instead check the batch size and poll every 100ms (up to 1s) until the
minimum batch size is accumulated.

vyzo committed:
store the remote trace peer address in the peerstore

that way we don't have to connect every time we open the stream.

vyzo committed:
make tracer.Close safer

Stebalien commented on tracer.go:
```diff
@@ -160,12 +173,15 @@ type RemoteTracer struct {
-	tr := &RemoteTracer{ctx: ctx, host: host, pi: pi, basicTracer: basicTracer{ch: make(chan struct{}, 1), lossy: true}}
+	tr := &RemoteTracer{ctx: ctx, host: host, peer: pi.ID, basicTracer: basicTracer{ch: make(chan struct{}, 1), lossy: true}}
+	for _, addr := range pi.Addrs {
+		host.Peerstore().AddAddr(pi.ID, addr, peerstore.PermanentAddrTTL)
```
`host.Peerstore().AddAddrs(pi.ID, pi.Addrs, ...)`?

Stebalien replied:
(we can do them all at once)

vyzo replied:
ah, good point; will do.

Stebalien commented on tracer.go:
```diff
@@ -187,15 +203,17 @@ func (t *RemoteTracer) doWrite() {
+			t.mx.Unlock()
+			time.Sleep(100 * time.Millisecond)
+			goto again
+		}
```
for loop?

```go
t.mx.Lock()
for len(t.buf) < MinTraceBatchSize && time.Now().Before(deadline) { 
  t.mx.Unlock()
  time.Sleep(100 * time.Millisecond)
  t.mx.Lock()
}
```

vyzo replied:
the alergy to gotos strikes again! Ok, I'll write it as a for loop.

Stebalien reviewed (approved):
nits but LGTM

vyzo committed:
nits and beauty

raulk commented:
Hey, would’ve liked to review this properly before it was merged to master :-( Didn’t get a chance due to other things, but I’d have appreciated a mention and a heads-up warning that the merge train was leaving.

vyzo commented:
@raulk you were extremely busy, so didn't want to burden you.
Feel free to make comments about potential improvements, and we can address them.
Note that the grpc trace client will live out of tree, as we don't want a dependency on grpc for mainline.
